### DISCUSSION TITLE
Cannot create and then update the same record in a projector because of the $exists property in Model class

### DISCUSSION THREAD

**Discussion by KhoaDuongUQ** [spatie/laravel-event-sourcing#488] [Q&A]
Hello all, first post here!

For my side project, I have 2 tables games & players, and I have a GameProjector to listen to the GameCreatedEvent like below:

```
class GameProjector extends Projector
{
    public function onGameCreated(GameCreatedEvent $event)
    {
        $game = new Game([
            'id' => $event->gameId,
            'created_at' => $event->createdAt(),
            'updated_at' => $event->createdAt(),
        ]);

        $game->writeable()->save();

        event(new PlayerJoinedEvent(
            gameId: $event->gameId,
            playerId: $event->playerId,
            playerName: $event->playerName,
        ));
        
        // The record never gets updated here
        $game->writeable()->update([
            'host_id' => $event->playerId,
        ]);
    }
}
```

These tables have a circular foreign key constraints host_id and game_id, so that's why I need to dispatch the PlayerJoinedEvent event to insert the player record before I update the host_id value.

I have figured out the reason is because the `exists` property of Laravel Model never gets updated to true so calling `update` on the model object does nothing (line 1009 in Model class source code). The workaround can be `Game::find($event->gameId)->writable()->update(...)` but this feels still dirty because of an additional DB query.

Does anyone have any idea how to solve this? Thanks!

**Summary**: `$model->writeable()->save()` does not update the `exists` property of the $model object to true, so subsequent `update()` calls cannot update the record.

**Comment by inmanturbo** [comment]
Is your player joined game event a storable event? If so calling it here will cause duplicates if the events ever get replayed. Presumably you have a PlayerProjector that will write to the database onPlayerJoined.

You need a PlayerReactor (or GameReactor depending on how you want to structure your dependencies/domains) weighted with a 1 or higher so it runs after  the game projector. Give it an onGameCreated method. In there you can record your PlayerJoinedEvent.

Then your third piece of logic, the part that wasn't working, should be moved to a onPlayerJoined method in your game projector which takes in your PlayerJoinedEvent. Unless I'm misunderstanding your dependency chain. As for the other thing, you could try calling fresh() on the instance before attempting to update.

**Reply by KhoaDuongUQ** [reply]
Thanks for the reply! And yes, I do have a PlayerProjector to save the player into db. Here's the flow of the app:

![Screenshot 2024-10-04 at 9 17 27 AM](https://github.com/user-attachments/assets/0435f538-edc3-4c38-b6e3-3f6b8222599e)

At first I was thinking that I can drop both games & players table, and replay both GameProjector and PlayerProjector to get the correct results. Having a Reactor means I could have lost all the host players when the tables are deleted and the events are replayed?

For the third piece of logic, if I move that to the PlayerJoinedEvent, I will need to retrieve the game, and only update the host_id if host_id is NULL, is it correct?

**Reply by inmanturbo** [reply]
You can still keep the projector, you use the reactor for things you want to happen only the first time an event occurs. If you have a dependency where you need to trigger other events, you want it in the reactor and not the projector.

Because you want to record the second event during "game play" the first time it happens, that way you have the data. But replaying events should never cause more events to get recorded. You just replay existing ones.
